### PR TITLE
Allow manual crafting of many atmos devices with the crafting menu

### PR DESCRIPTION
**Pull request by Ghilker**
## About The Pull Request
Adds a way to manually create basic atmos components with the crafting menu
Not all of the RPD available devices can be crafted, just the most basic ones.
Those are: Pipes, Color adapter, Layer adapter, Portable connector, H/E pipe, H/E junction, Heat exchangers, pressure pump, manual valve, filter, mixer, vent, scrubber, outlet injector, passive vent.
The available devices ensure solid setups for all your needs.

## Why It's Good For The Game
Allow hand crafting of atmos devices

## Changelog

:cl:
expansion: add the manual crafting of many atmos devices by the crafting menu
/:cl:

### CODE REVIEW

**Comment by wesoda25** [comment]
Could you please work on your PR names I can't tell what half of them do without reading the body

**Comment by Ghilker** [comment]
> Could you please work on your PR names I can't tell what half of them do without reading the body

hope this is better, i'm not good at naming things lol

**Comment by GuillaumePrata** [comment]
I know this is a request to help move forward with the construction rework, but having layer manifolds (and a few others) there just opens up some super easy sabotage by linking waste and distro.
The rework to smart pipes made it be annoying enough that you don't see people doing this as often and linking waste with distro in a random maint room will just end any round, it's not something that is viable to fix.

Having a traitor do irreversible damage to the station by just walking into maint with some metal kinda goes against the idea of making construction more efficient so people just don't abandon the shift the second something explodes.

**Comment by optimumtact** [comment]
as we all know antags can't use RPD's

**Comment by GuillaumePrata** [comment]
> as we all know antags can't use RPD's

Something something antag power creep something then since at least the RPD takes some effort steal or materials to print, metal is everywhere.

File: code/datums/components/crafting/recipes.dm [hunk 1]
```diff
@@ -1326,5 +1326,363 @@
...
+	reqs = list(/obj/item/stack/sheet/iron = 1)
+	time = 0.5 SECONDS
+	category = CAT_ATMOSPHERIC
+
+/datum/crafting_recipe/pipe/on_craft_completion(mob/user, atom/result)
+	var/obj/item/pipe/crafted_pipe = result
+	crafted_pipe.pipe_type = /obj/machinery/atmospherics/pipe/smart
+	crafted_pipe.pipe_color = COLOR_VERY_LIGHT_GRAY
+	crafted_pipe.p_init_dir = ALL_CARDINALS
+	crafted_pipe.setDir(SOUTH)
```
Comment (ATH1909) [inline code/datums/components/crafting/recipes.dm:1342]: would it be possible to make this match the direction the crafter is facing?

File: code/game/machinery/pipe/construction.dm [hunk 2]
```diff
@@ -299,6 +299,27 @@ Buildable meters
 		C.blood_volume = 0
 	return(OXYLOSS|BRUTELOSS)
 
+/obj/item/pipe/examine(mob/user)
+	. = ..()
+	. += span_notice("The pipe layer is set to [piping_layer].")
+	. += span_notice("You can change the pipe layer by Alt-Clicking the device.")
+	. += span_notice("You can rotate the device by using it in hand.")
```
Comment (ATH1909) [inline code/game/machinery/pipe/construction.dm:306]: "the device"?

**Comment by Pickle-Coding** [comment]
> The rework to smart pipes made it be annoying enough that you don't see people doing this as often and linking waste with distro in a random maint room

Pretty sure smart pipes made that process significantly easier. Place grey layer adapter, and now they're connected.

File: code/datums/components/crafting/recipes.dm (hunk 1 above)
Comment (Ghilker) [reply code/datums/components/crafting/recipes.dm:1342]: there is no need for that for smart pipes, but i'll do that for the other components

File: code/game/machinery/pipe/construction.dm (hunk 2 above)
Comment (Ghilker) [reply code/game/machinery/pipe/construction.dm:306]: dunno how else call them, since they are atmospheric devices

**Comment by GuillaumePrata** [comment]
> > The rework to smart pipes made it be annoying enough that you don't see people doing this as often and linking waste with distro in a random maint room
> 
> Pretty sure smart pipes made that process significantly easier. Place grey layer adapter, and now they're connected.

You still need a RPD to do that, before you just unwrenched 2 pipes in maint and linked both since they weren't color coded.

File: code/datums/components/crafting/recipes.dm [hunk 3]
```diff
@@ -1326,5 +1326,363 @@
...
+		/obj/item/pipe = 1,
+		/obj/item/stack/sheet/iron = 1)
+	time = 1 SECONDS
+	category = CAT_ATMOSPHERIC
+
+/datum/crafting_recipe/layer_adapter/check_requirements(mob/user, list/collected_requirements)
+	var/obj/item/pipe/required_pipe = collected_requirements[/obj/item/pipe][1]
+	if(ispath(required_pipe.pipe_type, /obj/machinery/atmospherics/pipe/smart))
+		return TRUE
+	to_chat(user, span_boldwarning("You can't craft a [name] without a pipe fitting!"))
```
Comment (Mothblocks) [inline code/datums/components/crafting/recipes.dm:1359]:
```suggestion
	to_chat(user, span_boldwarning("You can't craft \a [name] without a pipe fitting!"))
```

File: code/datums/components/crafting/recipes.dm [hunk 4]
```diff
@@ -1326,5 +1326,363 @@
...
+		/obj/item/pipe = 1,
+		/obj/item/stack/sheet/iron = 1)
+	time = 1 SECONDS
+	category = CAT_ATMOSPHERIC
+
+/datum/crafting_recipe/color_adapter/check_requirements(mob/user, list/collected_requirements)
+	var/obj/item/pipe/required_pipe = collected_requirements[/obj/item/pipe][1]
+	if(ispath(required_pipe.pipe_type, /obj/machinery/atmospherics/pipe/smart))
+		return TRUE
+	to_chat(user, span_boldwarning("You can't craft a [name] without a pipe fitting!"))
```
Comment (Mothblocks) [inline code/datums/components/crafting/recipes.dm:1383]:
```suggestion
	to_chat(user, span_boldwarning("You can't craft \a [name] without a pipe fitting!"))
```

File: code/datums/components/crafting/recipes.dm [hunk 5]
```diff
@@ -1326,5 +1326,363 @@
...
+		/obj/item/pipe = 1,
+		/obj/item/stack/sheet/iron = 1)
+	time = 1 SECONDS
+	category = CAT_ATMOSPHERIC
+
+/datum/crafting_recipe/he_pipe/check_requirements(mob/user, list/collected_requirements)
+	var/obj/item/pipe/required_pipe = collected_requirements[/obj/item/pipe][1]
+	if(ispath(required_pipe.pipe_type, /obj/machinery/atmospherics/pipe/smart))
+		return TRUE
+	to_chat(user, span_boldwarning("You can't craft a [name] without a pipe fitting!"))
```
Comment (Mothblocks) [inline code/datums/components/crafting/recipes.dm:1407]:
```suggestion
	to_chat(user, span_boldwarning("You can't craft \a [name] without a pipe fitting!"))
```

File: code/datums/components/crafting/recipes.dm [hunk 6]
```diff
@@ -1326,5 +1326,363 @@
...
+		/obj/item/pipe = 1,
+		/obj/item/stack/sheet/iron = 1)
+	time = 1 SECONDS
+	category = CAT_ATMOSPHERIC
+
+/datum/crafting_recipe/he_junction/check_requirements(mob/user, list/collected_requirements)
+	var/obj/item/pipe/required_pipe = collected_requirements[/obj/item/pipe][1]
+	if(ispath(required_pipe.pipe_type, /obj/machinery/atmospherics/pipe/smart))
+		return TRUE
+	to_chat(user, span_boldwarning("You can't craft a [name] without a pipe fitting!"))
```
Comment (Mothblocks) [inline code/datums/components/crafting/recipes.dm:1431]:
```suggestion
	to_chat(user, span_boldwarning("You can't craft \a [name] without a pipe fitting!"))
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
I'm going to tire myself out replacing all those "a"'s, please just make it a proc that they all call.

File: code/game/machinery/pipe/construction.dm (hunk 2 above)
Comment (Mothblocks) [reply code/game/machinery/pipe/construction.dm:306]: "rotate it"?

Comment (Mothblocks) [reply code/game/machinery/pipe/construction.dm:306]: How about just "rotate it"? Or are you referring to like, accessories

Comment (Ghilker) [reply code/game/machinery/pipe/construction.dm]: i'll go with rotate it

File: code/datums/components/crafting/recipes.dm [hunk 7]
```diff
@@ -45,6 +45,14 @@
 /datum/crafting_recipe/proc/on_craft_completion(mob/user, atom/result)
 	return
 
+///Check if the pipe used for atmospheric device crafting is the proper one
+/datum/crafting_recipe/proc/atmos_pipe_check(mob/user, list/collected_requirements)
```
Comment (Razharas) [inline code/datums/components/crafting/recipes.dm:49]: a good idea would have been to make a /pipe/ subtype and move all pipes and this proc there like food stuff does

Comment (Mothblocks) [reply code/datums/components/crafting/recipes.dm:49]: long time no see

Comment (Mothblocks) [reply code/datums/components/crafting/recipes.dm:49]: (youre right though)

**Pull request merged**